cdn_lambda/functions/map_to_s3.py

```diff
@@ -0,0 +1,61 @@
+import json
+import logging
+import os
+from datetime import datetime
+
+import boto3
+
+LOG = logging.getLogger("map-to-s3-lambda")
+
+
+def lambda_handler(event, context):
+    _ = context
+
+    # This AWS Lambda function must be deployed with DB_TABLE_NAME and
+    # DB_TABLE_REGION environment variables
+    table_name = os.getenv("DB_TABLE_NAME")
+    table_region = os.getenv("DB_TABLE_REGION")
+
+    # Get uri from origin request event
+    request = event["Records"][0]["cf"]["request"]
+    web_uri = request["uri"]
+
+    # Query latest item with matching uri from DynamoDB table
+    LOG.info("Querying '%s' table for '%s'...", table_name, web_uri)
+
+    query_result = boto3.client("dynamodb", region_name=table_region).query(
+        TableName=table_name,
+        Limit=1,
+        ScanIndexForward=False,
+        KeyConditionExpression="web_uri = :u and from_date <= :d",
+        ExpressionAttributeValues={
+            ":u": {"S": web_uri},
+            ":d": {"S": str(datetime.utcnow())},
+        },
+    )
+
+    if query_result["Items"]:
+        LOG.info("Item found for '%s'", web_uri)
+
+        # Update request uri to point to S3 object key
+        latest_item = query_result["Items"][0]
+
+        try:
+            request["uri"] = latest_item["object_key"]["S"]
+            return request
+        except KeyError:
```

Comment:
A few points here:

1. The possible case I mentioned earlier where object_key could be missing was just an example - in reality, there's unlimited possible issues, right?  For instance, the query to DynamoDB could fail with a connection error.  If we do have to implement our own logging for errors, it seems like it needs to surround everything which could fail.  I was thinking more like a `@decorator` which can easily catch everything, rather than adding try/except blocks.

2. I don't think it's a good idea to try to make our own error message in this way - I mean, just because you got a KeyError, you don't really *know* that 'object_key' was the missing key.  And the KeyError exception itself formats to a string telling you which key was missing.  So, adding this kind of message duplicates info which is already available but makes it possibly inaccurate (as you might have guessed wrong why a KeyError was raised).

3. Should we explicitly return 500 actually, or just let the exception propagate?  I haven't investigated what happens if an exception is allowed to propagate from a CloudFront lambda, but it might already be giving us some nice features for free - like generating a unique ID for the failure which can be cross-referenced in other logs.  This is something missing from your return 500.  Also, you're allowing the client to see the entire content of the DynamoDB record, we probably shouldn't allow that in general, we might be stashing metadata in there that we don't want customers to see.

So putting that all together:

- If we are responsible for logging our own exceptions, then the best way is probably to wrap the function with a decorator, which calls LOG.exception("Internal error during request").  LOG.exception logs the full backtrace.
- I suggest that decorator can also re-raise the exception after logging, rather than explicitly returning 500, and then we should have a look what happens in CloudFront in that case.
- If you think it's important for the latest_item to be logged in case of error, that can be included via try/except *as well as* the decorator, for instance:

```
latest_item = query_result["Items"][0]
try:
  ... # code using latest_item
except:
  LOG.exception("Exception occurred while processing %s", latest_item)
  raise
```

Reply:
So, I've broadened the except and logged a more generic exception like you've mentioned, raising the caught exception. I'll experiment on this a bit and see if we need to do any other handling.